module/VuFind/src/VuFind/ILS/Driver/PAIA.php

```diff
@@ -1774,6 +1787,43 @@ protected function paiaPostAsArray($file, $data)
      */
     protected function paiaLogin($username, $password)
     {
+        // as PAIA supports two authentication methods (defined as grant_type:
+        // password or client_credentials), check which one is configured
+        if (!in_array($this->grantType, ['password', 'client_credentials'])) {
+            throw new ILSException(
+                'Unsupported PAIA grantType configured.'
+            );
+        }
+
+        // prepare http header
+        $header_data = [];
+
+        // prepare post data depending on configured grant type
+        switch ($this->grantType) {
+        case 'password':
+            $post_data = [
+                "username"   => $username,
+                "password"   => $password
+            ];
+            break;
+        case 'client_credentials':
+            // client_credentials only works if we have client_credentials
+            // username and password
+            if (isset($this->config['PAIA']['client_username'])
+                && isset($this->config['PAIA']['client_password'])
```

Comment:
Should you have an else that throws an exception here in case of misconfiguration?

Reply:
Good point, added an else and an explicit exception message.

Reply:
Looks like this new message also uses grant_type instead of grantType. I think this is a sign that it may be worthwhile at some point to clean up the inconsistent naming conventions within this class. Generally, VuFind uses camelCase, but this code has quite a bit of snake_case in it.

Reply:
Thanks for the clarification, @olli-gold -- I hadn't realized that some of this terminology was coming directly from PAIA. In that case, I suspect that most of my feedback can be safely ignored. :-)

Reply:
Thanks @olli-gold exactly that's the case. So far I opted for camelCase whenever it is VuFind related (variables, configuration) but stick with PAIA syntax in documentation and configuration. The only divergence from this rule I am aware of are the configuration values `client_password`and `client_username`. Would changing those to `clientPassword`and `clientUsername`add to more consitency? Or would `client_credentialPassword` and `client_credentialUsername` be even more explicit and consistent?

Reply:
I think clientPassword and clientUsername might be slightly preferable to client_password and client_username in the name of consistency... but based on my other comment, maybe we'll end up renaming them entirely. We should probably make sure we're all on the same page about intent before we finalize naming. (For example, if we want to be able to override user-entered credentials regardless of login method, maybe these become overrideUsername / overridePassword).

Reply:
I refactored the variable names in the latest commit.